### PR TITLE
feat(backend/blocks): enhance list concatenation with advanced operations

### LINKED ISSUE #11139: Implement block to concatenate lists
We need a block that concatenates two or more lists into one

### PR DESCRIPTION
**Pull request by aviu16**
## Summary

Enhances the existing `ConcatenateListsBlock` and adds five new companion blocks for comprehensive list manipulation, addressing issue #11139 ("Implement block to concatenate lists").

### Changes

- **Enhanced `ConcatenateListsBlock`** with optional deduplication (`deduplicate`) and None-value filtering (`remove_none`), plus an output `length` field
- **New `FlattenListBlock`**: Recursively flattens nested list structures with configurable `max_depth`
- **New `InterleaveListsBlock`**: Round-robin interleaving of elements from multiple lists
- **New `ZipListsBlock`**: Zips corresponding elements from multiple lists with support for padding to longest or truncating to shortest
- **New `ListDifferenceBlock`**: Computes set difference between two lists (regular or symmetric)
- **New `ListIntersectionBlock`**: Finds common elements between two lists, preserving order

### Helper Utilities

Extracted reusable helper functions for validation, flattening, deduplication, interleaving, chunking, and statistics computation to support the blocks and enable future reuse.

### Test Coverage

Comprehensive test suite with 188 test functions across 29 test classes covering:
- Built-in block test harness validation for all 6 blocks
- Manual edge-case tests for each block (empty inputs, large lists, mixed types, nested structures)
- Internal method tests for all block classes
- Unit tests for all helper utility functions

Closes #11139

## Test plan

- [x] All files pass Python syntax validation (`ast.parse`)
- [x] Built-in `test_input`/`test_output` tests defined for all blocks
- [x] Manual tests cover edge cases: empty lists, large lists, mixed types, nested structures, deduplication, None removal
- [x] Helper function tests validate all utility functions independently
- [x] All block IDs are valid UUID4
- [x] Block categories set to `BlockCategory.BASIC` for consistency with existing list blocks

<h2>Greptile Overview</h2>

<details><summary><h3>Greptile Summary</h3></summary>

Enhanced `ConcatenateListsBlock` with deduplication and None-filtering options, and added five new list manipulation blocks (`FlattenListBlock`, `InterleaveListsBlock`, `ZipListsBlock`, `ListDifferenceBlock`, `ListIntersectionBlock`) with comprehensive helper functions and test coverage.

**Key Changes:**
- Enhanced `ConcatenateListsBlock` with `deduplicate` and `remove_none` options, plus `length` output field
- Added `FlattenListBlock` for recursively flattening nested lists with configurable `max_depth`
- Added `InterleaveListsBlock` for round-robin element interleaving
- Added `ZipListsBlock` with support for padding/truncation
- Added `ListDifferenceBlock` and `ListIntersectionBlock` for set operations
- Extracted 12 reusable helper functions for validation, flattening, deduplication, etc.
- Comprehensive test suite with 188 test functions covering edge cases

**Minor Issues:**
- Helper function `_deduplicate_list` has redundant logic in the `else` branch that duplicates the `if` branch
- Three helper functions (`_filter_empty_collections`, `_compute_list_statistics`, `_chunk_list`) are defined but unused - consider removing unless planned for future use
- The `_make_hashable` function uses `hash(repr(item))` for unhashable types, which correctly treats structurally identical dicts/lists as duplicates
</details>

<details><summary><h3>Confidence Score: 4/5</h3></summary>

- Safe to merge with minor style improvements recommended
- The implementation is well-structured with comprehensive test coverage (188 tests), proper error handling, and follows existing block patterns. All blocks use valid UUID4 IDs and correct categories. The helper functions provide good code reuse. The minor issues are purely stylistic (redundant code, unused helpers) and don't affect functionality or safety.
- No files require special attention - both files are well-tested and follow project conventions
</details>

<details><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant User
    participant Block as List Block
    participant Helper as Helper Functions
    participant Output
    
    User->>Block: Input (lists/parameters)
    Block->>Helper: _validate_all_lists()
    Helper-->>Block: validation result
    
    alt validation fails
        Block->>Output: error message
    else validation succeeds
        Block->>Helper: _concatenate_lists_simple() / _flatten_nested_list() / etc.
        Helper-->>Block: processed result
        
        opt deduplicate enabled
            Block->>Helper: _deduplicate_list()
            Helper-->>Block: deduplicated result
        end
        
        opt remove_none enabled
            Block->>Helper: _filter_none_values()
            Helper-->>Block: filtered result
        end
        
        Block->>Output: result + length
    end
    
    Output-->>User: Block outputs
```
</details>

<sub>Last reviewed commit: a6d5445</sub>

<sub>(2/5) Greptile learns from your feedback when you react with thumbs up/down!</sub>

### CODE REVIEW

File: autogpt_platform/backend/backend/blocks/data_manipulation.py [hunk 1]
```diff
@@ -682,25 +682,251 @@ async def run(self, input_data: Input, **kwargs) -> BlockOutput:
...
+    result: List[Any],
+    current_depth: int,
+    max_depth: int,
+) -> None:
+    """Internal recursive helper for flattening nested lists."""
+    for item in items:
+        if isinstance(item, list) and (max_depth == -1 or current_depth < max_depth):
+            _flatten_recursive(item, result, current_depth + 1, max_depth)
+        else:
+            result.append(item)
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/blocks/data_manipulation.py:757]: @claude this will get caught by the executor and just error this one graph correct?

File: autogpt_platform/backend/backend/blocks/data_manipulation.py [hunk 2]
```diff
@@ -682,25 +682,251 @@ async def run(self, input_data: Input, **kwargs) -> BlockOutput:
...
+    else:
+        # For non-order-preserving, use a set-like approach
+        seen_set: set[int] = set()
+        result_list: List[Any] = []
+        for item in items:
+            item_id = _make_hashable(item)
+            if item_id not in seen_set:
+                seen_set.add(item_id)
+                result_list.append(item)
+        return result_list
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/blocks/data_manipulation.py]: Then why have two branches

File: autogpt_platform/backend/backend/blocks/data_manipulation.py [hunk 3]
```diff
@@ -682,25 +682,251 @@ async def run(self, input_data: Input, **kwargs) -> BlockOutput:
...
+def _make_hashable(item: Any) -> int:
+    """
+    Create a hashable representation of any item for deduplication.
+    Uses id-based hashing for unhashable types like dicts and lists.
+    """
+    try:
+        return hash(item)
+    except TypeError:
+        # For unhashable types (dict, list), use repr-based hashing
+        return hash(repr(item))
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/blocks/data_manipulation.py]: Is this an issue or just letting me know?

File: autogpt_platform/backend/backend/blocks/data_manipulation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aviu16) [reply autogpt_platform/backend/backend/blocks/data_manipulation.py:757]: Yes, the executor catches exceptions from block execution. I've also added a max recursion depth guard (1000 levels) as an additional safety measure to prevent stack overflow.

File: autogpt_platform/backend/backend/blocks/data_manipulation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aviu16) [reply autogpt_platform/backend/backend/blocks/data_manipulation.py]: Good catch! Simplified to a single approach. The preserve_order parameter was unnecessary since both branches were functionally identical.

File: autogpt_platform/backend/backend/blocks/data_manipulation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aviu16) [reply autogpt_platform/backend/backend/blocks/data_manipulation.py:757]: Yes, exactly right. The RecursionError raised at line 749 will bubble up through the block's run method and be caught by the executor, which will set an error status on just this block (and any downstream dependents). Other blocks in the graph that don't depend on this one will continue executing normally.

File: autogpt_platform/backend/backend/blocks/data_manipulation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aviu16) [reply autogpt_platform/backend/backend/blocks/data_manipulation.py]: Good question — the two branches handle structurally different types:

- **dict**: converted to a sorted tuple of (key, value) pairs so that `{"a": 1, "b": 2}` and `{"b": 2, "a": 1}` produce the same hash. Sorting is needed for deterministic comparison.
- **list/tuple**: converted to a plain tuple preserving order, since `[1, 2]` and `[2, 1]` should NOT match.

They share the recursion pattern but the structural transformation (sorted vs ordered) differs, so they need separate branches.

File: autogpt_platform/backend/backend/blocks/data_manipulation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aviu16) [reply autogpt_platform/backend/backend/blocks/data_manipulation.py]: Just letting you know — `_filter_none_values` is a straightforward utility used by `ConcatenateListsBlock.run()` to strip None entries before concatenation (when the user enables the `remove_none` option). No issue here, just documenting the helper for clarity.

**Comment by aviu16** [comment]
recheck

**Comment by aviu16** [comment]
Addressed the remaining review feedback in c1227253f:

1. **`_compute_nesting_depth` recursion guard** — Added `max_depth` parameter (defaults to `_MAX_FLATTEN_DEPTH=1000`) to prevent stack overflow on deeply nested inputs. This was correctly identified as being called before `_flatten_recursive` in `FlattenListBlock.run()`, so the guard is now consistent across both functions.

2. **`_make_hashable` TypeError with mixed-type dict keys** — `sorted()` now uses `key=lambda x: (str(type(x[0])), str(x[0]))` to handle dicts with mixed-type keys (e.g., `{1: "one", "two": 2}`) without crashing.

3. **`ZipListsBlock` None handling** — Both `_zip_truncate` and `_zip_pad` now filter out `None` values before processing, consistent with how `InterleaveListsBlock`/`_interleave_lists` handles them.

Added tests for all three edge cases.

**Comment by Otto-AGPT** [comment]
## Review Summary

**CLA:** ✅ Signed

### Addressed Issues (👍)
- ✅ Recursion guard added ()
- ✅ `_interleave_lists` handles all-None inputs
- ✅ `_make_hashable` handles mixed-type dict keys via `str(type())` sorting
- ✅ `_compute_nesting_depth` uses iterative stack (no recursion limit)
- ✅ `ZipListsBlock` filters None values
- ✅ Type annotation fixed (`seen: set` not `set[int]`)
- ✅ Unused helper functions never added (clean)

### Still Needs Work (👎)
- ❌ **Documentation placeholders** — `<!-- MANUAL: how_it_works -->` and `<!-- MANUAL: use_case -->` still present in `docs/integrations/block-integrations/basic.md` for:
  - Flatten List
  - Interleave Lists
  - List Difference  
  - List Intersection
  - Zip Lists

**Action needed:** Fill in the "How it works" explanations and use cases, or remove the placeholder comments if docs will be added separately.

---
Code is solid — just needs the doc placeholders resolved.

**Comment by aviu16** [comment]
hey, those MANUAL comments are actually just structural markers that wrap the content, not empty placeholders. each section has the how_it_works and use_case text filled in between the opening and closing tags, e.g.:

```
<!-- MANUAL: how_it_works -->
This block takes elements from each input list in round-robin order...
<!-- END MANUAL -->
```

its the same pattern used for all other blocks in basic.md. all 5 sections (Flatten List, Interleave Lists, List Difference, List Intersection, Zip Lists) have their docs written

**Comment by Otto-AGPT** [comment]
**Correction:** The documentation IS filled in — the CodeRabbit comments about placeholders were from an earlier state of the PR.

All issues are now addressed:
- ✅ All code issues fixed
- ✅ Documentation filled in for all 5 new blocks
- ✅ CLA signed

This PR is ready to merge. 🚀

**Comment by majdyz** [comment]
@aviu16 sorry, those are coming from AI, reviewing your change now :)
Thank you for your contribution ✨
